File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/git/GitLogParserStrategyTest.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser.git;

import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogParserStrategy;
import de.maibornwolff.codecharta.importer.scmlogparser.parser.ParserStrategyContractTest;
import de.maibornwolff.codecharta.model.input.Modification;
import org.assertj.core.util.Lists;
import org.junit.Before;
import org.junit.Test;

import java.util.List;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;

public class GitLogParserStrategyTest extends ParserStrategyContractTest {

    private static final List<String> FULL_COMMIT = asList(
            "commit ca1fe2ba3be4",
            "Author: TheAuthor <mail@example.com>",
            "Date:   Tue May 9 19:57:57 2017 +0200",
            "    the commit message",
            "M src/Main.java",
            "M src/Main.java",
            "M src/Util.java");

    private GitLogParserStrategy parserStrategy;

    @Before
    public void setup() {
        parserStrategy = new GitLogParserStrategy();
    }

    @Override
    protected List<String> getFullCommit() {
        return FULL_COMMIT;
    }

    @Override
    protected LogParserStrategy getLogParserStrategy() {
        return parserStrategy;
    }

    @Override
    protected Stream<String> getTwoCommitsAsStraem() {
        List<String> twoCommits = Lists.newArrayList("commit");
        twoCommits.addAll(FULL_COMMIT);
        twoCommits.add("commit abcdef");
        twoCommits.addAll(FULL_COMMIT);
        twoCommits.add("commit");
        return twoCommits.stream();
    }

    @Test
    public void parsesFilenameFromFileMetadata() {
        String fileMetadata = "M\t src/Main.java";
        Modification modification = parserStrategy.parseModification(fileMetadata);
        assertThat(modification.getFilename()).isEqualTo("src/Main.java");
        assertThat(modification.getType()).isEqualTo(Modification.Type.MODIFY);
    }

    @Test
    public void parsesFilenameFromFileMetadataWithRename() {
        String fileMetadata = "R094\t srcs/Main.java\t src/Main.java";
        Modification modification = parserStrategy.parseModification(fileMetadata);
        assertThat(modification.getFilename()).isEqualTo("src/Main.java");
        assertThat(modification.getType()).isEqualTo(Modification.Type.RENAME);
    }

    @Test
    public void parsesFilenamesFromUnusualFileMetadata() {
        assertThat(parserStrategy.parseModification("")).isEqualTo(Modification.EMPTY);
        assertThat(parserStrategy.parseModification("  src/Main.java").getFilename())
                .isEqualTo("src/Main.java");
    }

    @Test
    public void parsesAuthorWithoutEmail() {
        String author = parserStrategy.parseAuthor("Author: TheAuthor");
        assertThat(author).isEqualTo("TheAuthor");
    }

    @Test
    public void parsesAuthorFromAuthorAndEmail() {
        String author = parserStrategy.parseAuthor("Author: TheAuthor <mail@example.com>");
        assertThat(author).isEqualTo("TheAuthor");
    }

}

File: analysis/import/SCMLogParser/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/git/GitLogParserStrategyTest.kt
package de.maibornwolff.codecharta.importer.scmlogparser.parser.git

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogParserStrategy
import de.maibornwolff.codecharta.importer.scmlogparser.parser.ParserStrategyContractTest
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.util.Lists
import org.junit.Test
import java.util.Arrays.asList
import java.util.stream.Stream

class GitLogParserStrategyTest : ParserStrategyContractTest() {

    private var parserStrategy: GitLogParserStrategy = GitLogParserStrategy()

    override val fullCommit: List<String>
        get() = FULL_COMMIT

    override val logParserStrategy: LogParserStrategy
        get() = parserStrategy

    override val twoCommitsAsStraem: Stream<String>
        get() {
            val twoCommits = Lists.newArrayList("commit")
            twoCommits.addAll(FULL_COMMIT)
            twoCommits.add("commit abcdef")
            twoCommits.addAll(FULL_COMMIT)
            twoCommits.add("commit")
            return twoCommits.stream()
        }

    @Test
    fun parsesFilenameFromFileMetadata() {
        val fileMetadata = "M\t src/Main.java"
        val modification = GitLogParserStrategy.parseModification(fileMetadata)
        assertThat(modification.currentFilename).isEqualTo("src/Main.java")
        assertThat(modification.type).isEqualTo(Modification.Type.MODIFY)
    }

    @Test
    fun parsesFilenameFromFileMetadataWithRename() {
        val fileMetadata = "R094\t srcs/Main.java\t src/Main.java"
        val modification = GitLogParserStrategy.parseModification(fileMetadata)
        assertThat(modification.currentFilename).isEqualTo("src/Main.java")
        assertThat(modification.oldFilename).isEqualTo("srcs/Main.java")
        assertThat(modification.type).isEqualTo(Modification.Type.RENAME)
    }

    @Test
    fun parsesFilenameFromAddedFile() {
        val fileMetadata = "A\t src/Main.java"
        val modification = GitLogParserStrategy.parseModification(fileMetadata)
        assertThat(modification.currentFilename).isEqualTo("src/Main.java")
        assertThat(modification.type).isEqualTo(Modification.Type.ADD)
    }

    @Test
    fun parsesFilenameFromDeletedFile() {
        val fileMetadata = "D\t src/Main.java"
        val modification = GitLogParserStrategy.parseModification(fileMetadata)
        assertThat(modification.currentFilename).isEqualTo("src/Main.java")
        assertThat(modification.type).isEqualTo(Modification.Type.DELETE)
    }

    @Test
    fun parsesFilenamesFromUnusualFileMetadata() {
        assertThat(GitLogParserStrategy.parseModification("")).isEqualTo(Modification.EMPTY)
        assertThat(GitLogParserStrategy.parseModification("  src/Main.java").currentFilename)
                .isEqualTo("src/Main.java")
    }

    companion object {

        private val FULL_COMMIT = asList(
                "commit ca1fe2ba3be4",
                "Author: TheAuthor <mail@example.com>",
                "Date:   Tue May 9 19:57:57 2017 +0200",
                "    the commit message",
                "A src/Added.java",
                "M src/Modified.java",
                "D src/Deleted.java")
    }
}
